pkg/db/db_releases.go

```diff
@@ -0,0 +1,681 @@
+	span, ctx := tracer.StartSpanFromContext(ctx, "DBUpdateOrCreateRelease")
+	defer span.Finish()
+	retrievedRelease, err := h.DBSelectReleaseByVersion(ctx, transaction, release.App, release.ReleaseNumber, false)
+	if err != nil {
+		return err
+	}
+	if retrievedRelease == nil {
+		err = h.insertReleaseRow(ctx, transaction, release)
+		if err != nil {
+			return err
+		}
+	} else {
+		err = h.updateReleaseRow(ctx, transaction, release)
+		if err != nil {
+			return err
+		}
+	}
+	err = h.insertReleaseHistoryRow(ctx, transaction, release, false)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *DBHandler) DBDeleteFromReleases(ctx context.Context, transaction *sql.Tx, application string, releaseToDelete uint64) error {
+	span, ctx := tracer.StartSpanFromContext(ctx, "DBDeleteFromReleases")
+	defer span.Finish()
+
+	targetRelease, err := h.DBSelectReleaseByVersion(ctx, transaction, application, releaseToDelete, true)
+	if err != nil {
+		return err
+	}
+	if targetRelease == nil {
+		return nil
+	}
+	err = h.deleteReleaseRow(ctx, transaction, *targetRelease)
+	if err != nil {
+		return err
+	}
+	err = h.insertReleaseHistoryRow(ctx, transaction, *targetRelease, true)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *DBHandler) DBClearReleases(ctx context.Context, transaction *sql.Tx, application string) error {
+	span, ctx := tracer.StartSpanFromContext(ctx, "DBClearReleases")
+	defer span.Finish()
+
+	allReleases, err := h.DBSelectAllReleasesOfApp(ctx, transaction, application)
+	if err != nil {
+		return err
+	}
+	if allReleases == nil {
+		logger.FromContext(ctx).Sugar().Infof("App %s does not contain any releases. No action taken", application)
+		return nil
+	}
+	for _, releaseToDelete := range allReleases {
+		err = h.DBDeleteFromReleases(ctx, transaction, application, uint64(releaseToDelete))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// actual changes in tables
+
+func (h *DBHandler) deleteReleaseRow(ctx context.Context, transaction *sql.Tx, release DBReleaseWithMetaData) error {
+	deleteQuery := h.AdaptQuery(`DELETE FROM releases WHERE appname=? AND releaseversion=?`)
+	span, _ := tracer.StartSpanFromContext(ctx, "deleteReleaseRow")
+	defer span.Finish()
+	span.SetTag("query", deleteQuery)
+	_, err := transaction.Exec(
+		deleteQuery,
+		release.App,
+		release.ReleaseNumber,
+	)
+	if err != nil {
+		return fmt.Errorf(
+			"could not delete release for app '%s' and version '%v' from DB. Error: %w\n",
+			release.App,
+			release.ReleaseNumber,
+			err)
+	}
+	return nil
+}
+
+func (h *DBHandler) updateReleaseRow(ctx context.Context, transaction *sql.Tx, release DBReleaseWithMetaData) error {
+	insertQuery := h.AdaptQuery(`
+		UPDATE releases SET created=?, manifests=?, metadata=?, environments=?
+		WHERE appname=? AND releaseVersion=?;`)
+	span, ctx := tracer.StartSpanFromContext(ctx, "updateReleaseRow")
```

Comment:
creating the span should always be the first line in a function.

Reply:
Done